Algorithms/Graph/BellmanFord.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+using System.Collections.Generic;
+using DataStructures.Graph;
+
+namespace Algorithms.Graph
+{
+    /// <summary>
+    /// Bellman-Ford algorithm on directed weighted graph.
+    /// </summary>
+    /// <typeparam name="T">Generic type of data in the graph.</typeparam>
+    public class BellmanFord<T>
+    {
+        private readonly DirectedWeightedGraph<T> graph;
+        private readonly Dictionary<Vertex<T>, double> distances;
+        private readonly Dictionary<Vertex<T>, Vertex<T>?> predecessors;
+
+        public BellmanFord(DirectedWeightedGraph<T> graph)
+        {
+            this.graph = graph;
+            distances = new Dictionary<Vertex<T>, double>();
+            predecessors = new Dictionary<Vertex<T>, Vertex<T>?>();
+        }
+
+        /// <summary>
+        /// Runs the Bellman-Ford algorithm to find the shortest distances from the source vertex to all other vertices.
+        /// </summary>
+        /// <param name="sourceVertex">Source vertex for shortest path calculation.</param>
+        /// <returns>
+        /// A dictionary containing the shortest distances from the source vertex to all other vertices.
+        /// If a vertex is unreachable from the source, it will have a value of double.PositiveInfinity.
+        /// </returns>
+        public Dictionary<Vertex<T>, double> Run(Vertex<T> sourceVertex)
+        {
+            InitializeDistances(sourceVertex);
+            RelaxEdges();
+            CheckForNegativeCycles();
+            return distances;
+        }
+
+        private void InitializeDistances(Vertex<T> sourceVertex)
+        {
+            foreach (var vertex in graph.Vertices)
+            {
+                if (vertex != null)
+                {
+                    distances[vertex] = double.PositiveInfinity;
+                    predecessors[vertex] = null;
+                }
+            }
+
+            distances[sourceVertex] = 0;
+        }
+
+        private void RelaxEdges()
+        {
+            int vertexCount = graph.Count;
+
+            for (int i = 0; i < vertexCount - 1; i++)
+            {
+                foreach (var vertex in graph.Vertices)
+                {
+                    if (vertex != null)
+                    {
+                        RelaxEdgesForVertex(vertex);
+                    }
+                }
+            }
+        }
+
+        private void RelaxEdgesForVertex(Vertex<T> u)
+        {
+            foreach (var neighbor in graph.GetNeighbors(u))
+            {
+                if (neighbor == null)
```

Comment:
The same question, can it be null? It seems odd to me to return null from `GetNeighbors`

Reply:
I have just checked the implementation of the graph class and it's odd, but your check seems valid